### PR TITLE
[SPARK-37209][YARN][TESTS] Fix `YarnShuffleIntegrationSuite` releated UTs when using `hadoop-3.2` profile without `assembly/target/scala-%s/jars`

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
`YarnShuffleIntegrationSuite`, `YarnShuffleAuthSuite` and `YarnShuffleAlternateNameConfigSuite` will failed when using `hadoop-3.2` profile without `assembly/target/scala-%s/jars`,  the fail reason is `java.lang.NoClassDefFoundError: breeze/linalg/Matrix`.

The above UTS can succeed when using `hadoop-2.7` profile without `assembly/target/scala-%s/jars` because `KryoSerializer.loadableSparkClasses` can workaroud when `Utils.isTesting` is true, but `Utils.isTesting` is false when using `hadoop-3.2` profile.

After investigated, I found that when `hadoop-2.7` profile is used, `SPARK_TESTING`  will be propagated to AM and Executor, but when `hadoop-3.2` profile is used, `SPARK_TESTING`  will not be propagated to AM and Executor.

In order to ensure the consistent behavior of using `hadoop-2.7` and ``hadoop-3.2``, this pr change to manually propagate `SPARK_TESTING` environment variable if it exists to ensure `Utils.isTesting` is true in above test scenario.

### Why are the changes needed?
Ensure `YarnShuffleIntegrationSuite` releated UTs can succeed when using `hadoop-3.2` profile without `assembly/target/scala-%s/jars`

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?

- Pass the Jenkins or GitHub Action

- Manual test `YarnShuffleIntegrationSuite`. `YarnShuffleAuthSuite` and `YarnShuffleAlternateNameConfigSuite`  can be verified in the same way.

Please ensure that the `assembly/target/scala-%s/jars` directory does not exist before executing the test command, we can clean up the whole project by executing follow command or clone a new local code repo.

1. run with `hadoop-3.2` profile

```
mvn clean install -Phadoop-3.2 -Pyarn -Dtest=none -DwildcardSuites=org.apache.spark.deploy.yarn.YarnShuffleIntegrationSuite
```

**Before**

```
YarnShuffleIntegrationSuite:
- external shuffle service *** FAILED ***
  FAILED did not equal FINISHED (stdout/stderr was not captured) (BaseYarnClusterSuite.scala:227)
Run completed in 48 seconds, 137 milliseconds.
Total number of tests run: 1
Suites: completed 2, aborted 0
Tests: succeeded 0, failed 1, canceled 0, ignored 0, pending 0
*** 1 TEST FAILED ***
```

Error stack as follows:

```
21/11/20 23:00:09.682 main ERROR Client: Application diagnostics message: User class threw exception: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 0.0 failed 4 times,
 most recent failure: Lost task 0.3 in stage 0.0 (TID 6) (localhost executor 1): java.lang.NoClassDefFoundError: breeze/linalg/Matrix
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:348)
        at org.apache.spark.util.Utils$.classForName(Utils.scala:216)
        at org.apache.spark.serializer.KryoSerializer$.$anonfun$loadableSparkClasses$1(KryoSerializer.scala:537)
        at scala.collection.immutable.List.flatMap(List.scala:366)
        at org.apache.spark.serializer.KryoSerializer$.loadableSparkClasses$lzycompute(KryoSerializer.scala:535)
        at org.apache.spark.serializer.KryoSerializer$.org$apache$spark$serializer$KryoSerializer$$loadableSparkClasses(KryoSerializer.scala:502)
        at org.apache.spark.serializer.KryoSerializer.newKryo(KryoSerializer.scala:226)
        at org.apache.spark.serializer.KryoSerializer$$anon$1.create(KryoSerializer.scala:102)
        at com.esotericsoftware.kryo.pool.KryoPoolQueueImpl.borrow(KryoPoolQueueImpl.java:48)
        at org.apache.spark.serializer.KryoSerializer$PoolWrapper.borrow(KryoSerializer.scala:109)
        at org.apache.spark.serializer.KryoSerializerInstance.borrowKryo(KryoSerializer.scala:346)
        at org.apache.spark.serializer.KryoSerializationStream.<init>(KryoSerializer.scala:266)
        at org.apache.spark.serializer.KryoSerializerInstance.serializeStream(KryoSerializer.scala:432)
        at org.apache.spark.shuffle.ShufflePartitionPairsWriter.open(ShufflePartitionPairsWriter.scala:76)
        at org.apache.spark.shuffle.ShufflePartitionPairsWriter.write(ShufflePartitionPairsWriter.scala:59)
        at org.apache.spark.util.collection.WritablePartitionedIterator.writeNext(WritablePartitionedPairCollection.scala:83)
        at org.apache.spark.util.collection.ExternalSorter.$anonfun$writePartitionedMapOutput$1(ExternalSorter.scala:772)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1468)
        at org.apache.spark.util.collection.ExternalSorter.writePartitionedMapOutput(ExternalSorter.scala:775)
        at org.apache.spark.shuffle.sort.SortShuffleWriter.write(SortShuffleWriter.scala:70)
        at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
        at org.apache.spark.scheduler.Task.run(Task.scala:136)
        at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:507)
        at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1468)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:510)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassNotFoundException: breeze.linalg.Matrix
        at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:419)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:352)
        ... 32 more
```

**After**
```
YarnShuffleIntegrationSuite:
- external shuffle service
Run completed in 35 seconds, 188 milliseconds.
Total number of tests run: 1
Suites: completed 2, aborted 0
Tests: succeeded 1, failed 0, canceled 0, ignored 0, pending 0
All tests passed.
```

2. run with `hadoop-2.7` profile

```
mvn clean install -Phadoop-2.7 -Pyarn -Dtest=none -DwildcardSuites=org.apache.spark.deploy.yarn.YarnShuffleIntegrationSuite
```

**Before**

```
YarnShuffleIntegrationSuite:
- external shuffle service
Run completed in 30 seconds, 828 milliseconds.
Total number of tests run: 1
Suites: completed 2, aborted 0
Tests: succeeded 1, failed 0, canceled 0, ignored 0, pending 0
All tests passed.
```

**After**
```
YarnShuffleIntegrationSuite:
- external shuffle service
Run completed in 30 seconds, 967 milliseconds.
Total number of tests run: 1
Suites: completed 2, aborted 0
Tests: succeeded 1, failed 0, canceled 0, ignored 0, pending 0
All tests passed.
```

### CODE REVIEW

**Comment by LuciferYang** [comment]
Wait for CI, and will add PR description tomorrow

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 1]
```diff
@@ -905,6 +905,10 @@ private[spark] class Client(
       env(ENV_DIST_CLASSPATH) = dcp
     }
 
+    sys.env.get(SPARK_TESTING).foreach { testing =>
+      env(SPARK_TESTING) = testing
+    }
+
```
Comment (xkrogen) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:911]: Maybe combine these two?

```suggestion
    Seq(ENV_DIST_CLASSPATH, SPARK_TESTING).foreach { envVar =>
      sys.env.get(envVar).foreach(value => env(envVar) = value)
    }
```

**Review by srowen — COMMENTED** [review]
Seems reasonable if it works

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 1 above)
Comment (LuciferYang) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:911]: looks good

Comment (LuciferYang) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: 59a618e fix this

**Comment by LuciferYang** [comment]
> Seems reasonable if it works

I manually verified this pr

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 2]
```diff
@@ -1703,7 +1705,6 @@ private[spark] class YarnClusterApplication extends SparkApplication {
     conf.remove(JARS)
     conf.remove(FILES)
     conf.remove(ARCHIVES)
-
```
Comment (jerqi) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:1706]: nit: unnecessary changes?

Comment (LuciferYang) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: thanks

**Comment by LuciferYang** [comment]
cc @srowen @HyukjinKwon @dongjoon-hyun

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Do we need to backport this or was it specific to master?

**Comment by LuciferYang** [comment]
All branches using `hadoop-3.2` profile by default should be required this fix, let me double check this and feedback later

**Comment by dongjoon-hyun** [comment]
cc @sunchao

**Comment by LuciferYang** [comment]
@srowen @dongjoon-hyun @sunchao It seems that this issue has been fixed by some PR company, let me investigate it

**Comment by LuciferYang** [comment]
The problem still exists, but the method I provided before can not reproduce stably. I need to find a way to reproduce stably

**Comment by LuciferYang** [comment]
```
mvn clean package -DskipTests -Phadoop-3.2 -Pyarn
rm -rf assembly/target/scala-2.12
mvn clean test -pl resource-managers/yarn -Phadoop-3.2 -Pyarn -Dtest=none -DwildcardSuites=org.apache.spark.deploy.yarn.YarnShuffleIntegrationSuite
```
Now above commands can reproduce stably this issue for  `-Phadoop-3.2`

**Comment by LuciferYang** [comment]
But use similar way to test with `-Phadoop-2.7` will trigger another problem

```
mvn clean package -DskipTests -Phadoop-2.7 -Pyarn
rm -rf assembly/target/scala-2.12
mvn clean test -pl resource-managers/yarn -Phadoop-2.7 -Pyarn -Dtest=none -DwildcardSuites=org.apache.spark.deploy.yarn.YarnShuffleIntegrationSuite

```

```
Discovery starting.
Discovery completed in 274 milliseconds.
Run starting. Expected test count is: 1
YarnShuffleIntegrationSuite:
*** RUN ABORTED ***
  java.lang.NoClassDefFoundError: org/apache/hadoop/shaded/com/google/inject/servlet/ServletModule
  at java.lang.ClassLoader.defineClass1(Native Method)
  at java.lang.ClassLoader.defineClass(ClassLoader.java:757)
  at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
  at java.net.URLClassLoader.defineClass(URLClassLoader.java:468)
  at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
  at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
  at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
  at java.security.AccessController.doPrivileged(Native Method)
  at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
  at java.lang.ClassLoader.loadClass(ClassLoader.java:419)
  ...
  Cause: java.lang.ClassNotFoundException: org.apache.hadoop.shaded.com.google.inject.servlet.ServletModule
  at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
  at java.lang.ClassLoader.loadClass(ClassLoader.java:419)
  at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
  at java.lang.ClassLoader.loadClass(ClassLoader.java:352)
  at java.lang.ClassLoader.defineClass1(Native Method)
  at java.lang.ClassLoader.defineClass(ClassLoader.java:757)
  at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
  at java.net.URLClassLoader.defineClass(URLClassLoader.java:468)
  at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
  at java.net.URLClassLoader$1.run(URLClassLoader.java:369)

```

**Comment by sunchao** [comment]
@LuciferYang What is the stacktrace from `hadoop-3.2` profile? is it the same as the one in the PR description? the `hadoop-2.7` stacktrace looks like is due the `hadoop-3.2` dependency somehow was incorrectly picked (since `org.apache.hadoop.shaded.*` should only be from 3.2 profile.)

**Comment by LuciferYang** [comment]
> @LuciferYang What is the stacktrace from hadoop-3.2 profile? is it the same as the one in the PR description? 

@sunchao Yes, it's same as the one in the PR description, I'm looking for a stable way to reproduce.

**Comment by LuciferYang** [comment]
>  the hadoop-2.7 stacktrace looks like is due the hadoop-3.2 dependency somehow was incorrectly picked (since org.apache.hadoop.shaded.* should only be from 3.2 profile.)


This seems to be another issue I stumbled upon

**Comment by LuciferYang** [comment]
@sunchao 

I think  run

```
mvn clean install -Phadoop-3.2 -Pyarn -Dtest=none -DwildcardSuites=org.apache.spark.deploy.yarn.YarnShuffleIntegrationSuite
```

can easy reproduce this issue and we should ensure that the `assembly/target/scala-%s/jars` directory does not exist before executing the test command, we can clean up the whole project by execute the test commad or clone a new local code repo, the error stack same as the one in the PR description:

```
21/11/20 23:00:09.682 main ERROR Client: Application diagnostics message: User class threw exception: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 0.0 failed 4 times,
 most recent failure: Lost task 0.3 in stage 0.0 (TID 6) (localhost executor 1): java.lang.NoClassDefFoundError: breeze/linalg/Matrix
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:348)
        at org.apache.spark.util.Utils$.classForName(Utils.scala:216)
        at org.apache.spark.serializer.KryoSerializer$.$anonfun$loadableSparkClasses$1(KryoSerializer.scala:537)
        at scala.collection.immutable.List.flatMap(List.scala:366)
        at org.apache.spark.serializer.KryoSerializer$.loadableSparkClasses$lzycompute(KryoSerializer.scala:535)
        at org.apache.spark.serializer.KryoSerializer$.org$apache$spark$serializer$KryoSerializer$$loadableSparkClasses(KryoSerializer.scala:502)
        at org.apache.spark.serializer.KryoSerializer.newKryo(KryoSerializer.scala:226)
        at org.apache.spark.serializer.KryoSerializer$$anon$1.create(KryoSerializer.scala:102)
        at com.esotericsoftware.kryo.pool.KryoPoolQueueImpl.borrow(KryoPoolQueueImpl.java:48)
        at org.apache.spark.serializer.KryoSerializer$PoolWrapper.borrow(KryoSerializer.scala:109)
        at org.apache.spark.serializer.KryoSerializerInstance.borrowKryo(KryoSerializer.scala:346)
        at org.apache.spark.serializer.KryoSerializationStream.<init>(KryoSerializer.scala:266)
        at org.apache.spark.serializer.KryoSerializerInstance.serializeStream(KryoSerializer.scala:432)
        at org.apache.spark.shuffle.ShufflePartitionPairsWriter.open(ShufflePartitionPairsWriter.scala:76)
        at org.apache.spark.shuffle.ShufflePartitionPairsWriter.write(ShufflePartitionPairsWriter.scala:59)
        at org.apache.spark.util.collection.WritablePartitionedIterator.writeNext(WritablePartitionedPairCollection.scala:83)
        at org.apache.spark.util.collection.ExternalSorter.$anonfun$writePartitionedMapOutput$1(ExternalSorter.scala:772)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1468)
        at org.apache.spark.util.collection.ExternalSorter.writePartitionedMapOutput(ExternalSorter.scala:775)
        at org.apache.spark.shuffle.sort.SortShuffleWriter.write(SortShuffleWriter.scala:70)
        at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
        at org.apache.spark.scheduler.Task.run(Task.scala:136)
        at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:507)
        at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1468)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:510)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassNotFoundException: breeze.linalg.Matrix
        at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:419)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:352)
        ... 32 more
```
and this pr can fix this issue.


On the other hand, run 

```
mvn clean install -Phadoop-2.7 -Pyarn -Dtest=none -DwildcardSuites=org.apache.spark.deploy.yarn.YarnShuffleIntegrationSuite
```

always successful

**Comment by LuciferYang** [comment]
@srowen

I tested the 

```
mvn clean install -Phadoop-3.2 -Pyarn -Dtest=none -DwildcardSuites=org.apache.spark.deploy.yarn.YarnShuffleIntegrationSuite
```

on `branch-3.2`, `branch-3.1` and `branch-3.0`. `branch-3.2` and `branch-3.1` failed for the same reason as the master

```
Caused by: java.lang.ClassNotFoundException: breeze.linalg.Matrix
```

but branch 3.0 was failed with different reason:

```
Caused by: java.lang.SecurityException: class "javax.servlet.http.HttpSessionIdListener"'s signer information does not match signer information of other classes in the same package
```

and I after cherry-pick the fix of SPARK-32954 to `branch 3.0`, `branch 3.0` will also fail for the same reason as other branches

**Comment by srowen** [comment]
Should we end up with this change and the servlet change backported to 3.0? would that seem to resolve the known issues?

**Comment by LuciferYang** [comment]
Yes, servlet issue in branch-3.0 should be a known issue. 

I also test current pr on `branch-3.2`, `branch-3.1` and `branch-3.0`, for `branch-3.2` and `branch-3.1`, the current pr is effective.

 But for `branch-3.0`, I think we need to further check what needs to be backported in addition to SPARK-32954 because when I applied the change of current pr to `branch-3.0`, the problem was not completely fixed.

**Comment by srowen** [comment]
Is this OK to merge by itself and back port to ... 3.0?

**Comment by LuciferYang** [comment]
> Is this OK to merge by itself and back port to ... 3.0?

Yes

**Comment by srowen** [comment]
Jenkins test this please

**Comment by srowen** [comment]
Merged to master/3.2/3.1/3.0

**Review by HyukjinKwon — COMMENTED** [review]
LGTM2

**Comment by LuciferYang** [comment]
thanks all

**Comment by dongjoon-hyun** [comment]
Thank you all. +1, late LGTM.
